WordPress/openverse
# Fix fetching and types on single result pages

~~This PR is based on #2584 and should be reviewed after it~~ 2584 is already merged

## Fixes



Fixes #2553 by @obulat

## Description


This is a large PR that fixes fetching logic for single result pages.

The changes here are based on the template changes added in #2584.

### Refactor fetching
The `singleResult` middleware sets the media `id` before navigating to the route. If the media item is available in another media store (so, if it was a part of the search results, the related media list or it is the global audio), `middleware` also sets the media as the `singleResultStore`'s `mainMedia`.

_Note: Our Nuxt app has two rendering modes. SSR: **Server** renders the single result page if you open the single result page directly, by going to the direct URL like https://openverse.org/image/f9384235-b72e-4f1e-9b05-e1b116262a29.  CSR: **Client** renders the page if you navigate to the single result page from another page, such as when you click on a search result, or one of the related media items, or even on the global audio track._

On the server, we fetch the media item and serve the fully-rendered single result page if the fetching is successful, or an error page if there's a fetching error.

On client navigation, we fetch the media item asynchronously, in the `useFetch` composable in the single result page component. Most of the time we will be only fetching the related media because the media data is already set in the `middleware`. The only exception to this is when you click on the homepage images: those do not set the media data, and we have to fetch both the media and the related media data to render the page.

### Single result store
I added getters for `audio`/`image` that return media detail of the correct type.
The `fetch` method now returns the media of the correct type or throws an error. ~~The error that we get when fetching is sent to Sentry as is. After sending the error to Sentry, we save the error to the store to use it in the error pages (which should be updated to show different messages on errors other than 404).~~ This pattern caused sending the errors to Sentry twice: in the store's `fetchMedia`'s `catch` clause and after the error is re-thrown. So, I refactored it: The error is caught in the `fetchMedia` method in the store. We extract the statusCode and the message to set as the `fetchState.fetchingError` value, and then re-throw the error. This error is caught in the `middleware` (on SSR) or in the component's `useFetch` method, and _then_ sent to Sentry. After that, we redirect to the Nuxt error page.

While working on #2618 I realized that we can avoid capturing error in Sentry in a better way. The store was refactored to return either the media item or null. Null is returned if there is an error, which is caught in the `fetch` function, which updates the store `fetchingError` and sends the error to Sentry. Then, the components handle a return of `null` as the error to show the error page.

### Templates
In addition to changes in #2584, this PR adds a skeleton loader until the item's thumbnail is loaded.
The related images component had both a Loading icon and a Loading skeleton when fetching - I removed the loading icon.
The server-rendered image report page is also simplified. It only needs to be rendered on the server and does not need to be navigated to client-side, and it should not be rendered if there is an error loading the `image`, so I used the `asyncData` there to ensure that `image` is always non-null.

## Testing Instructions

Test loading the image/audio single result pages: the Server-side rendered version (loading directly) and the client-side rendered version (navigating from the search results, from the related media section, global audio or the homepage image links).
If you throttle your browser, you should see the loading skeleton when you click on the homepage images because we need to fetch the image data. In other cases of client-side navigation, we already have the data and the thumbnail, so the loading should be instant.
You can also update the default timeout set in the `api-service` to cause timeout errors.
Another way I used to test errors on the single result page is to change the search result link's href using devtools, making it invalid. Then, if you click this link you should be redirected to the 404 page. If you open the Vue devtools on this page, you should also see the error status code and the error message as the error page prop. 

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
